plugins/inputs/whois/whois.go

```diff
@@ -0,0 +1,223 @@
+package whois
+
+import (
+	_ "embed"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/config"
+	"github.com/influxdata/telegraf/plugins/inputs"
+	"github.com/likexian/whois"
+	"github.com/likexian/whois-parser"
+)
+
+//go:embed sample.conf
+var sampleConfig string
+
+type Whois struct {
+	Domains            []string        `toml:"domains"`
+	Server             string          `toml:"server"`
+	Timeout            config.Duration `toml:"timeout"`
+	IncludeNameServers bool            `toml:"include_name_servers"`
+	Log                telegraf.Logger `toml:"-"`
+
+	Client         *whois.Client
+	WhoisLookup    func(client *whois.Client, domain, server string) (string, error)
+	ParseWhoisData func(raw string) (whoisparser.WhoisInfo, error)
+}
+
+func (*Whois) SampleConfig() string {
+	return sampleConfig
+}
+
+func (w *Whois) Gather(acc telegraf.Accumulator) error {
+	now := time.Now()
+
+	if w.Client == nil {
+		if err := w.Init(); err != nil {
+			return err
+		}
+	}
+
+	for _, domain := range w.Domains {
+		w.Log.Debugf("Fetching WHOIS data for: %s", domain)
+		w.Log.Debugf("Using WHOIS server: %s with timeout: %v", w.Server, w.Timeout)
+
+		// Fetch WHOIS raw data
+		rawWhois, err := w.WhoisLookup(w.Client, domain, w.Server)
+		if err != nil {
+			w.Log.Errorf("WHOIS query failed for %s: %v", domain, err)
+			acc.AddError(err)
+
+			// Always register a metric, even on failure
+			acc.AddFields("whois", map[string]interface{}{
+				"status": 0, // Mark failure
+			}, map[string]string{
+				"domain":        domain,
+				"domain_status": "UNKNOWN",
+			})
+			continue
+		}
+
+		// Parse WHOIS data using whois-parser
+		parsedWhois, err := w.ParseWhoisData(rawWhois)
+		if err != nil {
+			w.Log.Errorf("WHOIS parsing failed for %s: %v", domain, err)
+			acc.AddError(err)
+
+			// Always register a metric, even on failure
+			acc.AddFields("whois", map[string]interface{}{
+				"status": 0, // Mark failure
+			}, map[string]string{
+				"domain":        domain,
+				"domain_status": "UNKNOWN",
+			})
+			continue
+		}
+
+		w.Log.Debugf("Parsed WHOIS data for %s: %+v", domain, parsedWhois)
+
+		// Prevent nil pointer panic
+		if parsedWhois.Domain == nil {
+			w.Log.Warnf("No domain info found for %s", domain)
+
+			// Always register a metric, even on failure
+			acc.AddFields("whois", map[string]interface{}{
+				"status": 0, // Mark failure
+			}, map[string]string{
+				"domain":        domain,
+				"domain_status": "UNKNOWN",
+			})
```

Comment:
`domain_status` is a tag. Not sure of the additional `status` field would be needed/helpful.
```suggestion
			// Always register a metric, even on failure
			acc.AddFields("whois", map[string]interface{}{
				"status":        0, // Mark failure
				"domain_status": "UNKNOWN",
			}, map[string]string{
				"domain": domain,
			})
```

Reply:
The `status` will allow users to track more easily with integers if something was failed on the whois process. But this can also be unified into something like `status_code` and as is already done in `x509_cert` when using `verification_code`, we can make it using a integer instead of supporting `domain_status` and `status`. Please let me know WDYT

Reply:
Good idea to unify with other plugins and use `status_code` instead.

Reply:
Unified, and readme updated.